google/cloud/ndb/_transaction.py

```diff
@@ -23,6 +23,123 @@
 log = logging.getLogger(__name__)
 
 
+class _Propagation(object):
+    """This class aims to emulate the same behaviour as was provided by the old
+    Datastore RPC library.
+
+    https://cloud.google.com/appengine/docs/standard/python/ndb/functions#context_options
+
+    It provides limited support for transactions within transactions. It has a
+    single public method func:`handle_propagation`.
+    """
+
+    def __init__(self, propagation, join=None, *args, **kwargs):
+        """Args:
+        propagation (int): The desired `propagation` option, corresponding
+            to a class:`TransactionOptions` option.
+        join (:obj:`bool`, optional): If the provided join argument must be
+            changed to conform to the requested propagation option then a
+            warning will be emitted. If it is not provided, it will be set
+            according to the propagation option but no warning is emitted.
+        """
+
+        # Avoid circular import in Python 2.7
+        from google.cloud.ndb import context as context_module
+
+        propagation_options = context_module.TransactionOptions._PROPAGATION
+        if propagation is None or propagation in propagation_options:
+            self.propagation = propagation
+        else:
+            raise ValueError(
+                "Unexpected value for propagation. Got: {}. Expected one of: "
+                "{}".format(propagation, propagation_options)
+            )
+
+        propagation_names = context_module.TransactionOptions._INT_TO_NAME
+        self.propagation_name = propagation_names.get(self.propagation)
+
+        self.join = join
+        joinable_options = context_module.TransactionOptions._JOINABLE
+        self.joinable = propagation in joinable_options
+        super(_Propagation, self).__init__(*args, **kwargs)
```

Comment:
Unecessary to call `super().__init__()` when just subclassing `object`.

Reply:
1704ec3